pkg/build/reaper/reaper.go

```diff
@@ -0,0 +1,113 @@
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/kubectl"
+	kutilerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/util"
+)
+
+// NewBuildConfigReaper returns a new reaper for buildConfigs
+func NewBuildConfigReaper(oc *client.Client) kubectl.Reaper {
+	return &BuildConfigReaper{oc: oc, pollInterval: kubectl.Interval, timeout: kubectl.Timeout}
+}
+
+// BuildConfigReaper implements the Reaper interface for buildConfigs
+type BuildConfigReaper struct {
+	oc                    client.Interface
+	pollInterval, timeout time.Duration
+}
+
+// Stop deletes the build configuration and all of the associated builds.
+func (reaper *BuildConfigReaper) Stop(namespace, name string, timeout time.Duration, gracePeriod *kapi.DeleteOptions) (string, error) {
+	noBcFound := false
+	noBuildFound := true
+
+	// Add deletion pending annotation to the build config
+	for {
+		bc, err := reaper.oc.BuildConfigs(namespace).Get(name)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				noBcFound = true
+				break
+			}
+			return "", err
+		}
+		if _, ok := bc.Annotations[buildapi.BuildConfigDeletionAnnotation]; ok {
+			break
+		}
+
+		// Set deletion pending annotation
+		if err := util.AddObjectAnnotations(bc, map[string]string{buildapi.BuildConfigDeletionAnnotation: ""}); err != nil {
+			return "", err
+		}
+
+		_, err = reaper.oc.BuildConfigs(namespace).Update(bc)
+		if err == nil {
+			break
+		}
+		if !kerrors.IsConflict(err) {
+			return "", err
+		}
+	}
+
+	// Collect builds related to the config.
+	builds, err := reaper.oc.Builds(namespace).List(buildutil.BuildConfigSelector(name), nil)
+	if err != nil {
+		return "", err
+	}
+	errList := []error{}
+	for _, build := range builds.Items {
+		noBuildFound = false
+		if err := reaper.oc.Builds(namespace).Delete(build.Name); err != nil {
+			glog.Warningf("Cannot delete Build %s/%s: %v", build.Namespace, build.Name, err)
+			if !kerrors.IsNotFound(err) {
+				errList = append(errList, err)
+			}
+		}
+	}
+
+	// Collect deprecated builds related to the config.
+	// Note: Delete this block after BuildConfigLabelDeprecated is removed.
+	builds, err = reaper.oc.Builds(namespace).List(buildutil.BuildConfigSelectorDeprecated(name), nil)
+	if err != nil {
+		return "", err
+	}
+	for _, build := range builds.Items {
+		noBuildFound = false
+		if err := reaper.oc.Builds(namespace).Delete(build.Name); err != nil {
+			glog.Warningf("Cannot delete Build %s/%s: %v", build.Namespace, build.Name, err)
+			if !kerrors.IsNotFound(err) {
+				errList = append(errList, err)
+			}
+		}
+	}
+
+	// Aggregate all errors
+	if len(errList) > 0 {
+		glog.Warningf("Failed to delete all Builds associated with BuildConfig %s/%s. You can try re-running the command or removing them manually", namespace, name)
```

Comment:
"BuildConfig %s/%s will not be deleted because not all associated builds could be deleted.  You can try....."

(point being you need to tell them the buildconfig was not deleted)